File: python/src/eof_toolbox/eof.py
"""Grid-level EOF solver."""

from __future__ import annotations

from typing import Any

import numpy as np

from .core import PackedEof
from .reshape import GridPacker
from .weights import SpatialWeights, latitude_weights, make_weights


class Eof:
    """EOF analysis of a ``(n_lon, n_lat, n_time)`` field.

    The eigenproblem is solved at construction. Retrieve maps and PCs
    with :meth:`maps` / :meth:`pcs`, or reconstruct anomalies with
    :meth:`reconstruct`.

    Parameters
    ----------
    data :
        Spatiotemporal array. NaNs mark unused cells and must be
        constant in time.
    lat :
        Latitude in degrees, length ``n_lat``. Builds area weights
        ``sqrt(cos(lat))`` unless ``weights`` is given.
    n_eof :
        Number of modes. Default is ``min(n_valid_space, n_time)``.
    weights :
        Multiplied onto ``data`` before the eigenproblem.
    weighting :
        ``'sqrtcos'``, ``'cos'``, or ``'none'`` when ``lat`` is used.
    """

    def __init__(
        self,
        data: np.ndarray,
        lat: np.ndarray | None = None,
        n_eof: int | None = None,
        *,
        weights: np.ndarray | None = None,
        weighting: str = "sqrtcos",
    ) -> None:
        data = np.asarray(data, dtype=float)
        if data.ndim != 3:
            raise ValueError("data must be (n_lon, n_lat, n_time)")
        self.n_lon, self.n_lat, self.n_time = data.shape
        self.spatial_weights = SpatialWeights(
            self.n_lon, self.n_lat, lat=lat, weights=weights, method=weighting
        )
        self._packer = GridPacker(self.n_lon, self.n_lat)
        packed, self._loc = self._packer.pack(self.spatial_weights.apply(data))
        if packed.shape[0] == 0:
            raise ValueError("no finite spatial locations")
        self._solver = PackedEof(packed, n_eof=n_eof)

    @property
    def n_eof(self) -> int:
        return self._solver.n_eof

    @property
    def eigenvalues(self) -> np.ndarray:
        return self._solver.eigenvalues

    @property
    def expvar(self) -> np.ndarray:
        """Percent of total *weighted* variance."""
        return self._solver.expvar

    def maps(self, n_eof: int | None = None, *, unweight: bool = True,
             standardize: bool = False) -> np.ndarray:
        """Spatial patterns, shape ``(n_lon, n_lat, n_eof)``."""
        L = self._scaled_modes(n_eof, unweight=unweight, standardize=standardize)
        k = L.shape[1]
        return self._packer.unpack(L, self._loc)[:, :, :k]

    def pcs(self, n_eof: int | None = None, *, standardize: bool = False) -> np.ndarray:
        """Principal components, shape ``(n_eof, n_time)``."""
        k = self.n_eof if n_eof is None else int(n_eof)
        Y = self._solver.pcs[:k, :].copy()
        if standardize:
            std = np.sqrt(np.abs(self._solver.eigenvalues[:k]))
            std_safe = np.where(std > 0, std, 1.0)
            Y = Y / std_safe[:, np.newaxis]
        return Y

    def reconstruct(self, n_eof: int | None = None) -> np.ndarray:
        """Reconstructed anomalies in physical units, shape of the input."""
        maps = self.maps(n_eof, unweight=True, standardize=False)
        pcs = self.pcs(n_eof, standardize=False)
        weighted_maps = maps * self.spatial_weights.array[:, :, np.newaxis]
        return np.einsum("ijm,mt->ijt", weighted_maps, pcs)

    def _scaled_modes(
        self, n_eof: int | None, *, unweight: bool, standardize: bool
    ) -> np.ndarray:
        k = self.n_eof if n_eof is None else int(n_eof)
        L = self._solver.modes[:, :k].copy()
        if unweight:
            w = self.spatial_weights.flatten()[self._loc].astype(float).copy()
            w[w == 0] = np.inf
            L = L / w[:, np.newaxis]
            L[~np.isfinite(L)] = 0.0
        if standardize:
            std = np.sqrt(np.abs(self._solver.eigenvalues[:k]))
            std_safe = np.where(std > 0, std, 1.0)
            L = L * std_safe
        return L


def eof(
    data: np.ndarray,
    lat: np.ndarray | None = None,
    n_eof: int | None = None,
    *,
    weights: np.ndarray | None = None,
    weighting: str = "sqrtcos",
    unweight: bool = True,
    standardize: bool = False,
    **_: Any,
) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """One-shot wrapper around :class:`Eof` (MATLAB-compatible returns)."""
    solver = Eof(
        data, lat=lat, n_eof=n_eof, weights=weights, weighting=weighting
    )
    maps = solver.maps(unweight=unweight, standardize=standardize)
    pcs = solver.pcs(standardize=standardize)
    return maps, pcs, solver.expvar, solver.eigenvalues


__all__ = ["Eof", "eof", "latitude_weights", "make_weights", "SpatialWeights"]

File: matlab/eof.m
function [eof_maps, pcs, expvar, eig_values] = eof(data, varargin)
%EOF Empirical Orthogonal Functions of a (lon, lat, time) field.
%
%   [EOF_MAPS, PCS, EXPVAR, EIG_VALUES] = EOF(DATA, ...)
%
%   DATA is an n_lon x n_lat x n_time array (NaNs mark unused cells).
%   The time mean is removed. Spatial modes are returned on the same
%   horizontal grid; missing cells remain NaN.
%
%   Spatial (area / latitude) weighting
%   -----------------------------------
%   Equal-area grids do not need extra weights. On a regular lat-lon
%   grid, polar cells are smaller, so they must not count as much as
%   equatorial cells in the covariance. Pass LAT to apply the standard
%   weight sqrt(max(cosd(lat),0)) (NCAR Climate Data Guide).
%
%   After the eigenproblem, spatial maps are *unweighted* (divided by
%   W) so EOF_MAPS is in the physical units of DATA. Set 'unweight'
%   false to keep maps in weighted space (orthonormal columns).
%
%   Syntax
%     eof(data)
%     eof(data, lat)
%     eof(data, n_eof)
%     eof(data, lat, n_eof)
%     eof(data, lat, n_eof, 'std')
%     eof(data, 'lat', lat, 'n_eof', k)
%     eof(data, 'weights', W)                 % W multiplied as-is
%     eof(data, 'lat', lat, 'weighting', 'sqrtcos')  % default
%     eof(data, 'lat', lat, 'weighting', 'cos')      % legacy v1
%     eof(data, 'lat', lat, 'unweight', false)
%
%   Name/value options
%     'lat'        latitude vector, length n_lat, degrees
%     'weights'    (n_lat,), (n_lon,n_lat), or (n_lon*n_lat,) weights
%                  multiplied onto DATA before the eigenproblem
%     'weighting'  'sqrtcos' (default) | 'cos' | 'none'  (used with lat)
%     'n_eof'      number of modes (default: min(space, time))
%     'unweight'   divide maps by W after analysis (default true)
%     'std'        scale maps by sqrt(lambda) and PCs by 1/sqrt(lambda)
%
%   Outputs
%     EOF_MAPS     n_lon x n_lat x n_eof spatial patterns
%     PCS          n_eof x n_time principal components
%     EXPVAR       n_eof x 1 percent of total (weighted) variance
%     EIG_VALUES   n_eof x 1 covariance eigenvalues (variance per mode)
%
%   Reconstruction of anomalies (with default unweight true):
%     data_anom ≈ (eof_maps .* W) contracted with pcs, or equivalently
%     maps_weighted * pcs, where maps_weighted = eof_maps .* W.
%
%   See also MEOF, EOFCORE, EOF_MAKE_WEIGHTS.
%
%   Zelun Wu, Xiamen University & University of Delaware, 15 May 2020.
%   Area weighting (sqrt(cos lat)), unweighting, name/value API: 2026.

narginchk(1, inf)
[n_lon, n_lat, n_time] = size(data);
n_loc = n_lon * n_lat;
opt = eof_parse_options(varargin, n_lon, n_lat, n_time, n_loc);

W = eof_make_weights(n_lon, n_lat, opt.lat, opt.weights, opt.weighting);
W3 = repmat(W, [1, 1, n_time]);
data_w = data .* W3;

[data_2d, in_nonan_locations] = reshape3dto2d(data_w);
n_valid = size(data_2d, 1);
n_eof = min(opt.n_eof, min(n_valid, n_time));

[eof_2d, pcs, eig_values, expvar] = eofcore(data_2d, n_eof);

if opt.unweight
    w_flat = W(:);
    w_valid = w_flat(in_nonan_locations);
    w_valid(w_valid == 0) = Inf;  % pole weights: leave map at 0
    eof_2d = eof_2d ./ repmat(w_valid, [1, n_eof]);
    eof_2d(~isfinite(eof_2d)) = 0;
end

if opt.flag_std
    for n = 1:n_eof
        std_n = sqrt(abs(eig_values(n)));
        if std_n > 0
            eof_2d(:, n) = eof_2d(:, n) * std_n;
            pcs(n, :) = pcs(n, :) / std_n;
        end
    end
end

eof_maps = reshape2dto3d(eof_2d, [n_lon, n_lat, n_eof], in_nonan_locations);
end
